test apps package with multiple CPUs

wjordan opened the pull request:
Improves build/test times for apps-package artifacts:
- Use uglifyjs-webpack-plugin v1.3.0 with `parallel` option
- refactor test-low-memory.sh script to run tests using parallel/xargs even in non-circleci environments.
- small changes to Rakefile build/package scripts for clearer logging output

wjordan commented:
I think the riskiest part of this PR is the version-override for `uglifyjs-webpack-plugin`, which uses the newer (but already itself deprecated) [`uglify-es`](https://github.com/code-dot-org/code-dot-org/pull/26586/files#diff-1b28c39bb54275ba36a716e98f13b8cbR11989) package with our (older) version of webpack. I think as long as tests pass it should be fine, but I'd like second opinion from someone who knows what's going on in our js code internally a bit better than myself.

islemaster commented:
Reading through I don't see anything alarming, except for the number of dependencies added in yarn.lock just from pinning the uglify version.  When I pulled the change to try locally though I ran into some issues.  Not sure yet if this is a problem with the change or with my local setup, but might be an early warning for rollout problems to the rest of the team.

`yarn build:dist` absolutely clobbered my machine.  Chrome ran out of memory and crashed.  Running it in a terminal with little else going on, the build completed in about five minutes.  To compare I switched back to a recent staging commit and `build:dist` took ~10 minutes but didn't interfere with other work.  So maybe there's a question of where we expect to run this command (I rarely/never run `build:dist` on my local machine).

islemaster commented:
@davidsbailey Wondering if you can give this a shot on OSX and see how it performs for you.

davidsbailey commented:
on staging, `yarn build:dist` took 13m58s with no impact to system usability. on the branch in this PR, `yarn build:dist` took 7m33s with the system being somewhat slower at times but definitely still usable. My laptop is probably middle-of-the-road or perhaps on the older side relative to our team in terms of specs:
<img width="584" alt="screen shot 2019-01-21 at 11 08 50 pm" src="https://user-images.githubusercontent.com/8001765/51518060-9368cc80-1dd1-11e9-8ea0-62df6d03cf36.png">

davidsbailey commented:
@wjordan , I just want to over-communicate that I'm about to attempt to upgrade webpack to version 4, where `webpack.optimize.UglifyJsPlugin` appears to have been replaced by `config.optimization.minimize`  https://stackoverflow.com/questions/49570488/upgrade-to-webpack-4-uglifyjsplugin-error

You should feel free to proceed with this PR and I can try to preserve this functionality in the upgrade. However, if you were thinking that going straight to `config.optimization.minimize` sounded better for any reason, then you'd be welcome to wait until after the webpack upgrade is complete to pursue this.

islemaster commented:
Sounds like webpack 4's uglify operation is [parallel by default.](https://blog.box.com/blog/how-we-improved-webpack-build-performance-95)

wjordan commented:
Thanks @davidsbailey, I'll pause this PR in favor of the webpack 4 update if that will be ready soon, which should simplify the `uglifyjs-webpack-plugin` update and eliminate the need for pulling in that already-deprecated `uglify-es` package (the webpack4-compatible version of `uglifyjs-webpack-plugin` depends on a newer version of `uglify-js`, and as @islemaster mentioned the parallel-uglify setting will be enabled by default).

Once the webpack4 update is in, i'll rebase/resolve this PR so it will pull in only the parallel-test change.

wjordan commented:
@islemaster @davidsbailey Now that we've upgraded to Webpack 4 I've updated this PR to only include the changes to run the test script in parallel.

I added a few lines of bash / awk to check `MemFree` in `/proc/meminfo` to limit the number of parallel processes based on free memory on the system (which should hopefully prevent any crashes when running locally).

I'll probably manually test the changes out on `staging` / `test` soon, but I'd also like to know whether it runs okay locally on your machine(s) as well to make sure there are no hiccups

wjordan commented:
> I'll probably manually test the changes out on staging / test soon

Merging this PR now, since another check prevents me from manually testing these changes without committing the changes. Would still appreciate a review after the fact if possible, and I will revert quickly if this PR introduces any other issues after being merged.
